### DISCUSSION TITLE
Casadi MPC optimization cost function

### DISCUSSION THREAD

**Discussion by garyguo1998** [casadi/casadi#3296] [Q&A]
for i in range(0, num_horizon):
        # dynamics: pos
        constraint = vertcat(
            constraint,
            X[X_DIM * (i + 1) + X_ID["x"]]
            - (X[X_DIM * i + X_ID["x"]] + X[X_DIM * i + X_ID["v"]] * timestep),
        )
        # dynamics: spd
        constraint = vertcat(
            constraint,
            X[X_DIM * (i + 1) + X_ID["v"]]
            - (X[X_DIM * i + X_ID["v"]] + U[U_DIM * i + U_ID["acc"]] * timestep),
        )
        # constraint on battery output power, P_batt = P_drive + P_aux
        f_resistance = mu_tire * mass_veh * g
        spd_t = X[X_DIM * i + X_ID["v"]]
        acc_t = U[U_DIM * i + U_ID["acc"]]
        F_b = U[U_DIM * i + U_ID["F_b"]]
        P_drive = (
            p00
            + p01 * f_resistance / n_ratio
            + (p10 * n_ratio + p11 * f_resistance) * spd_t
            + (p01 * c_wind / n_ratio) * spd_t * spd_t
            + p11 * c_wind * (spd_t**3)
            + (p01 * mass_veh / n_ratio) * acc_t
            + (p01 / n_ratio) * F_b
            + p11 * mass_veh * spd_t * acc_t 
            + p11 * spd_t * F_b
        )
        constraint = vertcat(
            constraint,
            n_Ser * v_batt * U[U_DIM * i + U_ID["I"]]
            - U[U_DIM * i + U_ID["I"]]
            * U[U_DIM * i + U_ID["I"]]
            * (n_Ser / n_Par)
            * r_batt
            - P_aux
            - P_drive,
        )
        
        c=log(exp(U[U_DIM * i + U_ID["I"]])+exp(-U[U_DIM * i + U_ID["I"]]))/batt_param.Q_batt
        if i==0:
            delta_v=v_lv-spd_ego
            s_star = s_0+v_lv*t_gap+v_lv*delta_v/2/np.sqrt(acc*dec)
            vdot_lv = acc*(1-(v_lv/v_0)**delta - (s_star/s_lv)**2)
            airdrag1=c_wind*v_lv**2
            a=(-(f_resistance+airdrag1))/mass_veh
            l=1/(1+exp(-1000000*(vdot_lv-a)))
      
            Pdrive_lv=mass_veh*vdot_lv*v_lv+(f_resistance+airdrag1)*v_lv
            Pdrivelv1=l*Pdrive_lv
            s_lv=s_lv+(spd_ego-v_lv)*dt
            v_lv=v_lv+dt*vdot_lv
            
            
   
        else:
            delta_v = v_lv-spd_t
            s_star = s_0+v_lv*t_gap+v_lv*delta_v/2/np.sqrt(acc*dec)
            vdot_lv = acc*(1-(v_lv/v_0)**delta - (s_star/s_lv)**2)
            airdrag1=c_wind*v_lv**2
            a=(-(f_resistance+airdrag1))/mass_veh
            l=1/(1+exp(-1000000*(vdot_lv-a)))
            Pdrive_lv=mass_veh*vdot_lv*v_lv+(f_resistance+airdrag1)*v_lv
            Pdrivelv1=l*Pdrive_lv
            s_lv=s_lv+(spd_t-v_lv)*dt
            v_lv=v_lv+dt*vdot_lv
            
    
        # cost on minimizing the total energy, minimizing difference w.r.t. spd limit, minimizing acceleration
        cost = (
            cost
            +0.001*(P_drive)
            +0.0006*(Pdrivelv1)
            +20*((U[U_DIM*i+U_ID["acc"]])**2)
            +0.00000001*(age_p1*c**3+age_p2*c**2+age_p3*c)
            +1*(slack1[i]**2)
            +1*(slack2[i]**2)
           
        )
   
    # Terminal landing cost
    
    cost = (cost + 100*(X[X_DIM * (num_horizon) + X_ID["x"]] - final_pos_ego_pred)**2 
            + 100*(X[X_DIM * (num_horizon) + X_ID["v"]] - final_spd_ego_pred)**2
            )

The error is:
CasADi - 2023-07-23 16:21:11 WARNING("solver:nlp_grad_f failed: NaN detected for output grad_f_x, at (row 3, col 0).") [.../casadi/core/oracle_function.cpp:265]
CasADi - 2023-07-23 16:21:11 WARNING("solver:nlp_grad_f failed: NaN detected for output grad_f_x, at (row 3, col 0).") [.../casadi/core/oracle_function.cpp:265]

**Comment by garyguo1998** [comment]
I am not sure where is the error. The error may happen when I use the sigmoid function which is 
        a=(-(f_resistance+airdrag1))/mass_veh
        l=1/(1+exp(-1000000*(vdot_lv-a)))
        Pdrive_lv=mass_veh*vdot_lv*v_lv+(f_resistance+airdrag1)*v_lv
        Pdrivelv1=l*Pdrive_lv

**Comment by brunoml5** [comment]
Hi!

It's really hard to help if your code cannot be run in our PC. But try looking at your fourth variable (index 3) which appears in your objective. It might be envolved in a problematic expression such as division by zero or square-root of negative values.
